BUG: Fix regression on DataFrame.replace for regex

PetitLepton opened the pull request:
The commit ensures that the replacement for regex is not confined to the 
beginning of the string but spans all the characters within. The 
behaviour is then consistent with versions prior to 0.24.0.

- [x] closes #25259
- [x] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

jreback commented:
this seems highly related to https://github.com/pandas-dev/pandas/issues/24804 and a corresponding PR #24935 

can you have a look

jreback reviewed (changes requested):
changes *always* require a test & whatnews note, but see my comment on a related issue

charlesdong1991 commented:
i just took a quick look... seems the problems don't look the same... shall i move whatsnew in my PR to 0.24.2 from 0.25? @jreback

PetitLepton commented:
@jreback and @charlesdong1991 Hi, thanks for taking a look at this. I have very little knowledge about `pandas` internals but, from the description of #24804, it does not seem to be the same problem. I will try to include a test specifically for the case where the regex pattern is inside the string and not at the beginning. The current change already passes all the tests in `tests/frame/test_replace.py`.

pep8speaks commented:
Hello @PetitLepton! Thanks for updating the PR.



 - In the file [`pandas/tests/frame/test_replace.py`](https://github.com/pandas-dev/pandas/blob/1b05715eabe0def0f8af355fc02fab524624fa84/pandas/tests/frame/test_replace.py), following are the PEP8 issues :

> [Line 454:80](https://github.com/pandas-dev/pandas/blob/1b05715eabe0def0f8af355fc02fab524624fa84/pandas/tests/frame/test_replace.py#L454): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)

PetitLepton commented:
@jreback I added a test for the change and an entry in `whatsnew`. Please let me know if it matches `pandas` standard.

jreback commented on pandas/tests/frame/test_replace.py:
```diff
@@ -451,7 +451,11 @@ def test_regex_replace_dict_mixed(self):
         # nested dicts will not work until this is implemented for Series
-        mix = {'a': lrange(4), 'b': list('ab..'), 'c': ['a', 'b', np.nan, 'd']}
+        mix = {
+            'a': lrange(4),
```
can you preserve b from the original test and add another element

jreback replied:
add the issue number here as a comment as well

jreback commented:
ideally also replicate the test from the OP as well (this could be an alternative to changing the existing test)

WillAyd commented:
@PetitLepton can you address comments and merge master?

PetitLepton commented:
@WillAyd @jreback My apologies for the delay. I tried to follow the recommendations, reverting the initial test to its genuine form and added a simple test referring to the issue above.

WillAyd commented on pandas/tests/frame/test_replace.py:
```diff
@@ -466,6 +466,15 @@ def test_regex_replace_dict_nested(self):
         assert_frame_equal(res4, expec)
 
+    def test_regex_replace_dict_nested_non_first_character(self):
+        # Ensures that the regular expression is replaced
```
Minor but just leave a reference to the issue number in the comments, so `# GH 25259`

PetitLepton replied:
@WillAyd Do you mean that I remove the comment and replace it only with the issue number, namely `# GH 25259`?

WillAyd commented on doc/source/whatsnew/v0.24.2.rst:
```diff
@@ -23,6 +23,7 @@ Fixed Regressions
 - Fixed regression in :meth:`DataFrame.all` and :meth:`DataFrame.any` where ``bool_only=True`` was ignored (:issue:`25101`)
 - Fixed issue in ``DataFrame`` construction with passing a mixed list of mixed types could segfault. (:issue:`25075`)
 - Fixed regression in :meth:`DataFrame.apply` causing ``RecursionError`` when ``dict``-like classes were passed as argument. (:issue:`25196`)
+- Fixed regression in :meth:`DataFrame.replace` causing non-starting regex not being replaced. (:issue:`25259`)
```
```suggestion
- Fixed regression in :meth:`DataFrame.replace` where ``regex=True`` was only replacing patterns matching the start of the string (:issue:`25259`)
```

PetitLepton replied:
I am changing that now.

WillAyd reviewed (changes requested):
Minor nits otherwise lgtm

WillAyd commented on pandas/core/internals/managers.py:
```diff
@@ -1901,7 +1901,7 @@ def _compare_or_regex_match(a, b, regex=False):
     """
     Compare two array_like inputs of the same shape or two scalar values
```
Since we moved from `match` to `search` here can you also updated the function name?

PetitLepton replied:
OK.

PetitLepton committed:
BUG: Fix regression on DataFrame.replace for regex

The commit ensures that the replacement for regex is not confined to the 
beginning of the string but spans all the characters within. The 
behaviour is then consistent with versions prior to 0.24.0.

One test has been added to account for character replacement when the 
character is not at the beginning of the string.

PetitLepton commented:
Two tests are failing but I don't have access to the error messages.

WillAyd commented:
Hmm should have access but in any case here's the error:

```sh
##[error]./pandas/core/internals/managers.py(556,48): error E128: continuation line under-indented for visual indent
```

Also if you merge master instead of rebasing shouldn't need to force push which is preferable with maintaining GitHub comment history

PetitLepton committed:
Add space to match indentation

WillAyd reviewed (approved):
lgtm @jreback

TomAugspurger commented:
Thanks @PetitLepton!

PetitLepton commented:
@WillAyd and @jreback Thank you very much for your patience and guidance, it is my first (small) contribution to an open source project and I am really happy to have been able to help `pandas`. It is an amazing tool.

@TomAugspurger Thank you for approving.

supersational commented:
Just spent ages confused by this behaviour, thanks for fixing it @PetitLepton !

For those who need a quick-fix the following will work around the issue:
``` python3
# add this to the start of to_replace: r"(.*?)"
# add this to the start of replacement value: r"\1"
# e.g:
df = pd.DataFrame({'first': ['abc', 'bca', 'cab']})
df.replace({r"(.*?)"+'a':  r"\1"+'.'}, regex=True)

```
